### PR TITLE
Bluetooth: Mesh: Use dedicated thread to store mesh settings

### PR DESCRIPTION
**Pull request by PavelVPV**
This PR solves several issues:
- it solves an issue where the system workqueue can be blocked for significant time while the stack stores its configuration, which makes the device non-responsive until the storing is finished.
- it solves an issue where models that need to store model's data too frequently could wear out flash quickly.
- it solves an issue where models can block the BT RX thread or the system workqueue for significant time when large amount of data should be stored in the message opcode handler.

These issues are solved by adding the following features to the stack:
- adding a new thread and enabling it by default to process the settings store in a separate thread thus unblocking the system workqueue from processing other work items while the storing thread waits for the controller to allocate time to perform flash operations (write or erase). 
- adding an API that allows models to use the stack's work item to postpone storing of the model's data and use the dedicated thread to not block the BT RX thread or the system workqueue.

The complete description of the issues is described in the commits' messages.

### CODE REVIEW

**Comment by LingaoM** [comment]
There is a slight gap from what I thought, the current work has not been cleaned up completely, such as 

`bt_mesh_net_iv_update`  --> `bt_mesh_rpl_reset`  --> `clear_rpl`.

`node_reset_pending_handler` --> `bt_mesh_reset`  --> `bt_mesh_settings_store_pending`.

File: subsys/bluetooth/mesh/Kconfig [hunk 1]
```diff
@@ -1584,6 +1584,34 @@ config BT_MESH_SEQ_STORE_RATE
 	  off with a value that's guaranteed to be larger than the last
 	  one used before power off.
 
+config BT_MESH_SETTINGS_WORKQ
+	bool "Store the Bluetooth mesh settings in a separate work queue"
+	default y
+	help
+	  This option enables a separate cooperative thread which is used to
```
Comment (LingaoM) [inline subsys/bluetooth/mesh/Kconfig:1591]: As far as I know, once the cooperative thread is running, other threads cannot preempt it, so sysworkq and BT RX can only wait ?

Comment (LingaoM) [reply subsys/bluetooth/mesh/Kconfig:1591]: https://docs.zephyrproject.org/3.1.0/kernel/services/scheduling/index.html#cooperative-time-slicing

**Comment by PavelVPV** [comment]
> There is a slight gap from what I thought, the current work has not been cleaned up completely, such as
> 
> `bt_mesh_net_iv_update` --> `bt_mesh_rpl_reset` --> `clear_rpl`.
> 

True, but can be solved separately.

> `node_reset_pending_handler` --> `bt_mesh_reset` --> `bt_mesh_settings_store_pending`.
> 

Reset is not a concern here. The concern is to make the device responsive when it stores configuration. When the node is reset, it won't send or relay anything.

> `app_key_add` --> `bt_mesh_app_key_add` --> `update_app_key_settings`
> 

Only if the amount of changes in appkeys bigger than the `app_keys_update` array. Same for net keys and cdb. This is a corner case which probability is significantly low.

> It seems that there are still many storage operations directly carried out in BT RX, which clearly do not match the title of your PR

Right, I can rephrase that part in the documentation I added. But in general the said is truth. Most of the configuration is done in a separate work. There are more places that needs to be fixed and this will be done separately.

File: subsys/bluetooth/mesh/Kconfig (hunk 1 above)
Comment (PavelVPV) [reply subsys/bluetooth/mesh/Kconfig:1591]: They can, unless they enter a rescheduling point: https://docs.zephyrproject.org/latest/kernel/services/scheduling/index.html#concepts. You can observe the following stack call when storing settings:
`settings_save_one` -> `nvs_write`/`nvs_delete` -> `flash_write` -> `flash_nrf_write` (`soc_flash_nrf.c`) -> `write_synchronously` -> `nrf_flash_sync_exe` (`soc_flash_nrf_ticker.c`) -> `k_sem_take` (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/controller/flash/soc_flash_nrf_ticker.c#L225). At this point the thread is put on sleep until write or erase is finished. Other cooperative threads can be scheduled here.

**Comment by PavelVPV** [comment]
> https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/Kconfig#L131 In the previous scenario, both BT RX and Sysworkq may store Settings, and both have a considerable increase in Stack Size. If we put the storage operation in a separate task, should we also consider reducing BT RX and sysworkq stack size?

Not necessarily. BT RX and system workqueue are still used to for processing other mesh code. They are also used to process opcodes handlers (BT RX when a message is received over the air and system workqueue when the message is received via loopback) that go to the application code. Also, from what I saw, the stack size after provisioning is bigger than needed for the settings thread.

**Comment by Andrewpini** [comment]
Im curious. I can see that processing all settings related work in a separate queue will prevent starving other tasks in the system workqueue if there are multiple settings work items present in the system workqueue. But will it also prevent starvation in the case where a single settings work item is associated with huge delays, E.G. will the settings work queue yield while waiting for longer write/erase operations during handling of a settings work item, allowing the CPU to process items in the system workqueue simutaniously?

File: doc/connectivity/bluetooth/api/mesh/core.rst [hunk 2]
```diff
@@ -67,6 +67,46 @@ vulnerability and flash wear out.
...
+rescheduled until the work item is processed.
+
+When IV index, Sequence Number or CDB configuration should be stored, the work
+item is submitted to the workqueue without the delay. If the work item was
+previously scheduled, it will be rescheduled without the delay.
+
+The Replay Protection List uses the same work item to store RPL entries. If RPL
+entries store is requested and no other configuration is pending to be stored
+the delay will be set to :kconfig:option:`CONFIG_BT_MESH_RPL_STORE_TIMEOUT`.
+If other stack configuration should be stored, the delay defined by
```
Comment (Andrewpini) [inline doc/connectivity/bluetooth/api/mesh/core.rst:88]:
```suggestion
If other stack configuration should be stored, given that the delay defined by
```
?

File: doc/connectivity/bluetooth/api/mesh/core.rst [hunk 3]
```diff
@@ -67,6 +67,46 @@ vulnerability and flash wear out.
...
+:kconfig:option:`CONFIG_BT_MESH_RPL_STORE_TIMEOUT` and the work item was scheduled
+by the Replay Protection List, the work item will be rescheduled.
+
+When the work item is running, the stack will store all pending configuration
+including RPL entries.
+
+Work item execution context
+===========================
+
+The :kconfig:option:`CONFIG_BT_MESH_SETTINGS_WORKQ` configures context from which
```
Comment (Andrewpini) [inline doc/connectivity/bluetooth/api/mesh/core.rst:99]:
```suggestion
The :kconfig:option:`CONFIG_BT_MESH_SETTINGS_WORKQ` configures the context from which
```
?

File: doc/connectivity/bluetooth/api/mesh/core.rst [hunk 4]
```diff
@@ -67,6 +67,46 @@ vulnerability and flash wear out.
...
+the work item is executed. If the option is enabled, which is by default, the
+stack uses a dedicated cooperative thread to process the work item. This allows
+the stack to process other incoming and outgoing messages as well as other work
+items submitted to the system workqueue, while the stack configuration is being
+stored.
```
Comment (Andrewpini) [inline doc/connectivity/bluetooth/api/mesh/core.rst:104]:
```suggestion
the work item is executed. This option is enabled by default, and while enabled the
stack uses a dedicated cooperative thread to process the work item. This allows the
stack to handle persistent storage tasks without potentially blocking other system
workqueue tasks like incoming and outgoing messages.
```
?

File: doc/connectivity/bluetooth/api/mesh/core.rst [hunk 5]
```diff
@@ -67,6 +67,46 @@ vulnerability and flash wear out.
...
+
+When this option is disabled, the work item is submitted to the system workqueue.
+This means that the system workqueue is blocked for the time required to store
+the stack's configuration. It is not recommended to disable this option as this
+will make the device non-responsive for some noticeable time.
```
Comment (Andrewpini) [inline doc/connectivity/bluetooth/api/mesh/core.rst:109]:
```suggestion
the stack's configuration. It is not recommended to disable this option as this
will make the device non-responsive for a noticeable amount of time.
```
?

File: doc/connectivity/bluetooth/api/mesh/access.rst [hunk 6]
```diff
@@ -149,6 +149,22 @@ storage. The model can retrieve the data by calling the ``read_cb`` passed as
 a parameter to the callback. See the :ref:`settings_api` module documentation for
 details.
 
+When model data changes frequently, storing it on every change may speed up
+the wear of flash. To reduce the wear, the model can postpone the store by
+calling :c:func:`bt_mesh_model_data_store_schedule`. The stack will schedule a
+work with delay defined by :kconfig:option:`CONFIG_BT_MESH_STORE_TIMEOUT`. When
```
Comment (Andrewpini) [inline doc/connectivity/bluetooth/api/mesh/access.rst:155]:
```suggestion
work item with delay defined by :kconfig:option:`CONFIG_BT_MESH_STORE_TIMEOUT`. When
```
?

File: doc/connectivity/bluetooth/api/mesh/access.rst [hunk 7]
```diff
@@ -149,6 +149,22 @@ storage. The model can retrieve the data by calling the ``read_cb`` passed as
...
+the work is running, the stack will call the
+:c:member:`bt_mesh_model_cb.pending_store` callback for every model that
+requested the store. The model can then call :c:func:`bt_mesh_model_data_store`
+to store the data.
```
Comment (Andrewpini) [inline doc/connectivity/bluetooth/api/mesh/access.rst:159]:
```suggestion
:c:member:`bt_mesh_model_cb.pending_store` callback for every model that has
requested to store data. The model can then call :c:func:`bt_mesh_model_data_store`
to store the data.
```
?

**Review by Andrewpini — COMMENTED** [review]
The implementation looks good to me. Got a couple of comments to the docs.

**Comment by PavelVPV** [comment]
> Im curious. I can see that processing all settings related work in a separate queue will prevent starving other tasks in the system workqueue if there are multiple settings work items present in the system workqueue. But will it also prevent starvation in the case where a single settings work item is associated with huge delays, E.G. will the settings work queue yield while waiting for longer write/erase operations during handling of a settings work item, allowing the CPU to process items in the system workqueue simutaniously?

To clarify, there is only one settings work item regardless of amount of data that needs to be stored. Once this settings work is submitted, the mesh stack will go through all pending data that needs to be stored (see `store_pending` function in `mesh/settings.c`).

In short, yes, it is the reason of doing this. If the synchronization between radio and the flash driver is enabled, which is by default (see `CONFIG_SOC_FLASH_NRF_RADIO_SYNC_TICKER` for ZLL and `CONFIG_SOC_FLASH_NRF_RADIO_SYNC_MPSL` for SDC), the flash driver requests the controller for a chunk of time during which it can perform a flash operation. These chunks are fixed (https://github.com/zephyrproject-rtos/zephyr/blob/7b8170dd278ed40913d5df0944f34d81975ae8f0/drivers/flash/soc_flash_nrf.c#L40-L47) regardless of the amount of data that needs to be stored (or erased). From this point, the context (thread) that triggered the flash operation will sleep (https://github.com/zephyrproject-rtos/zephyr/blob/7b8170dd278ed40913d5df0944f34d81975ae8f0/subsys/bluetooth/controller/flash/soc_flash_nrf_ticker.c#L225) until all passed data is stored or all required flash pages are erased (unless some other failure happened). The chunks of time are not provided immediately. Until a chunk of time is provided, the rescheduling happens and wakes up other threads (including cooperative). When partial erase (`CONFIG_SOC_FLASH_NRF_PARTIAL_ERASE`) is enabled, there will be "gaps" between each erase cycle, which can be "occupied" by other threads.

Without this change, the rescheduling were still happening, but since the settings work item was submitted to the system workqueue, the stack operation was blocked because everything (except incoming messages) in stack works on the system workqueue.

File: subsys/bluetooth/mesh/settings.c [hunk 8]
```diff
@@ -40,6 +40,21 @@ LOG_MODULE_REGISTER(bt_mesh_settings);
...
+#define SETTINGS_WORKQ_PRIO 1
+#endif
+
+#ifdef CONFIG_BT_MESH_SETTINGS_WORKQ_STACK_SIZE
+#define SETTINGS_WORKQ_STACK_SIZE CONFIG_BT_MESH_SETTINGS_WORKQ_STACK_SIZE
+#else
+#define SETTINGS_WORKQ_STACK_SIZE 0
+#endif
+
+struct k_work_q settings_work_q;
```
Comment (alxelax) [inline subsys/bluetooth/mesh/settings.c:55]:
```suggestion
static struct k_work_q settings_work_q;
```

**Comment by alxelax** [comment]
LGTM just minor issue

**Review by alxelax — COMMENTED** [review]
LGTM just minor issue

File: subsys/bluetooth/mesh/Kconfig [hunk 9]
```diff
@@ -1584,6 +1584,34 @@ config BT_MESH_SEQ_STORE_RATE
...
+	  store Bluetooth mesh configuration. When this option is disabled,
+	  the stack's configuration is stored in the system workqueue. This
+	  means that the system workqueue will be blocked for the time needed
+	  to store the pending data. This time may significantly increase if
+	  the flash driver needs to erase few flash pages. Enabling this option
```
[the remark this answers is not in the document]
Comment (mia-ko) [reply subsys/bluetooth/mesh/Kconfig:1596]: Joining the party LOL
"a considerable amount of flash pages"?

File: subsys/bluetooth/mesh/Kconfig [hunk 10]
```diff
@@ -1584,6 +1584,34 @@ config BT_MESH_SEQ_STORE_RATE
...
+	  allows Bluetooth mesh to not block the system workqueue and thus
+	  process incoming and outgoing messages while the flash driver waits
+	  for the controller to allocate time needed to write data and/or erase
+	  required flash pages.
```
Comment (mia-ko) [inline subsys/bluetooth/mesh/Kconfig:1600]:
```suggestion
	  allows Bluetooth mesh not to block the system workqueue, and thus
	  process the incoming and outgoing messages while the flash driver waits
	  for the controller to allocate the time needed to write the data and/or erase
	  the required flash pages.
```

**Review by mia-ko — CHANGES_REQUESTED** [review]

**Comment by PavelVPV** [comment]
> `bt_mesh_net_iv_update` --> `bt_mesh_rpl_reset` --> `clear_rpl`.

Fix is here: https://github.com/zephyrproject-rtos/zephyr/pull/57317

File: doc/connectivity/bluetooth/api/mesh/access.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (PavelVPV) [reply doc/connectivity/bluetooth/api/mesh/access.rst:159]: I would leave `work is running`. Same wording used in here: https://docs.zephyrproject.org/latest/kernel/services/threads/workqueue.html

File: doc/connectivity/bluetooth/api/mesh/access.rst [hunk 11]
```diff
@@ -149,6 +149,22 @@ storage. The model can retrieve the data by calling the ``read_cb`` passed as
 a parameter to the callback. See the :ref:`settings_api` module documentation for
 details.
 
+When model data changes frequently, storing it on every change may lead to
+increased wear of flash. To reduce the wear, the model can postpone storing of
+data by calling :c:func:`bt_mesh_model_data_store_schedule`. The stack will
+schedule a work item with delay defined by the
+:kconfig:option:`CONFIG_BT_MESH_STORE_TIMEOUT` option. When the work item is
+running, the stack will call the :c:member:`bt_mesh_model_cb.pending_store`
+callback for every model that has requested to storing of data. The model can
```
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/access.rst:158]:
```suggestion
callback for every model that has requested storing of data. The model can
```

File: doc/connectivity/bluetooth/api/mesh/core.rst [hunk 12]
```diff
@@ -67,6 +67,46 @@ vulnerability and flash wear out.
...
+When the work item is running, the stack will store all pending configuration,
+including the RPL entries.
+
+Work item execution context
+===========================
+
+The :kconfig:option:`CONFIG_BT_MESH_SETTINGS_WORKQ` option configures the
+context from which the work item is executed. This option is enabled by
+default, and while enabled the stack uses a dedicated cooperative thread to
+process the work item. This allows the stack to process other incoming and
```
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/core.rst:102]:
```suggestion
default, and results in stack using a dedicated cooperative thread to
process the work item. This allows the stack to process other incoming and
```
?

**Review by mia-ko — COMMENTED** [review]
LGTM, just two small comments

**Review by mia-ko — DISMISSED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by jhedberg — DISMISSED** [review]

File: subsys/bluetooth/mesh/settings.c (hunk 8 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/settings.c:55]: It has not been fixed.

Comment (PavelVPV) [reply subsys/bluetooth/mesh/settings.c]: Sorry, though I fixed it.

**Review by mia-ko — APPROVED** [review]

**Review by alxelax — APPROVED** [review]

**Review by Andrewpini — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**